src/System.Private.CoreLib/src/System/Globalization/FormatProvider.FormatAndParse.Unix.cs

```diff
@@ -0,0 +1,164 @@
+                    return;
+                }
+                
+                byte* tempBuffer = stackalloc byte[ECVT_MAX_BUFFER_SIZE];
+                char* dst = number.digits;
+
+                number.scale = 0;
+                number.sign = false;
+                *dst = '\0';             
+
+                if (value < 0.0)
+                {
+                    number.sign = true;
+                }
+                
+                if (value == 0.0)
+                {
+                    for (int j=0; j<precision; j++)
+                    {
+                        dst[j] = '0';
+                    }
+                    dst[precision] = '\0';
+                    return;
+                }
+
+                //
+                // Get the number formatted as a string in the form x.xxxxxxexxxx
+                //
+
+                // "%.40e"
+                uint *format = stackalloc uint[2];
+                format[0] = 0x30342e25; // %.40
+                format[1] = 0x00000065; // e\0
+                int tempBufferLength = Interop.Sys.DoubleToString(value, (byte *) format, tempBuffer, ECVT_MAX_BUFFER_SIZE);
+                Debug.Assert(tempBufferLength>0 && ECVT_MAX_BUFFER_SIZE > tempBufferLength);
+
+                //
+                // Calculate the exponent value
+                //
+                
+                byte *pExponent = tempBuffer + tempBufferLength - 1;
+                while (*pExponent != 'e' && pExponent>tempBuffer)
+                {
+                    pExponent--;
+                }
+                
+                Debug.Assert(pExponent>tempBuffer && (pExponent < tempBuffer + tempBufferLength - 1));
+
+                int i = 1;
+                int exponentSign = 1;
+                if (pExponent[i] == '-')
+                {
+                    exponentSign = -1;
+                    i++;
+                }
+                else if (pExponent[i] == '+')
+                {
+                    i++;
+                }
+                
+                int exponentValue = 0;
+                while ((pExponent+i) < (tempBuffer + tempBufferLength)) 
+                {
+                    exponentValue = exponentValue * 10 + (pExponent[i] - (byte)'0');
+                    i++;
+                }
+                exponentValue *= exponentSign;
+
+                //
+                // Determine decimal location.
+                // 
+
+                if ( exponentValue == 0 )
+                {
+                    number.scale = 1;
+                }
+                else
+                {
+                    number.scale = exponentValue + 1;
+                }
+                
+                tempBufferLength = (int) (pExponent - tempBuffer); 
+
+                //
+                // Copy the string from the temp buffer upto precision characters, removing the sign, and decimal as required.
+                // 
+
+                i = 0;
+                byte *pTempBuffer = tempBuffer;
```

Comment:
This can also use better name - pTempBuffer and tempBuffer are very easy to confuse. It may be better to just use index instead - so that both source and destination are handled same way in the following loop.